File: backend/internal/tools/registry.go
package tools

import (
	"context"
	"fmt"
	"sync/atomic"

	"openaide/backend/internal/kernel"
)

type registryData struct {
	definitions map[string]kernel.ToolDefinition
	handlers    map[string]kernel.ToolHandler
}

// Registry is a lock-free tool registry using atomic.Value.
// Writes are rare (init only), reads are on every query.
type Registry struct {
	data             atomic.Value    // *registryData
	pendingQuestions chan string     // questions from ask_user tool (buffered, non-blocking)
}

// GetPendingQuestions drains and returns all pending user questions.
func (r *Registry) GetPendingQuestions() []string {
	if r.pendingQuestions == nil { return nil }
	var qs []string
	for {
		select {
		case q := <-r.pendingQuestions: qs = append(qs, q)
		default: return qs
		}
	}
}

// NewRegistry creates a lock-free tool registry.
func NewRegistry() *Registry {
	r := &Registry{pendingQuestions: make(chan string, 32)}
	r.data.Store(&registryData{
		definitions: make(map[string]kernel.ToolDefinition),
		handlers:    make(map[string]kernel.ToolHandler),
	})
	return r
}

func (r *Registry) load() *registryData { return r.data.Load().(*registryData) }

// Register adds a tool. Thread-safe but typically called during init.
func (r *Registry) Register(tool kernel.ToolDefinition, handler kernel.ToolHandler) error {
	name := tool.Function.Name
	if name == "" {
		return fmt.Errorf("tool name is empty")
	}

	old := r.load()
	if _, exists := old.definitions[name]; exists {
		return fmt.Errorf("tool already registered: %s", name)
	}

	// Copy-on-write
	newDefs := make(map[string]kernel.ToolDefinition, len(old.definitions)+1)
	newHandlers := make(map[string]kernel.ToolHandler, len(old.handlers)+1)
	for k, v := range old.definitions { newDefs[k] = v }
	for k, v := range old.handlers { newHandlers[k] = v }
	newDefs[name] = tool
	newHandlers[name] = handler
	r.data.Store(&registryData{definitions: newDefs, handlers: newHandlers})
	return nil
}

// Unregister removes a tool.
func (r *Registry) Unregister(name string) error {
	old := r.load()
	if _, exists := old.definitions[name]; !exists {
		return fmt.Errorf("tool not found: %s", name)
	}
	newDefs := make(map[string]kernel.ToolDefinition, len(old.definitions)-1)
	newHandlers := make(map[string]kernel.ToolHandler, len(old.handlers)-1)
	for k, v := range old.definitions {
		if k != name { newDefs[k] = v }
	}
	for k, v := range old.handlers {
		if k != name { newHandlers[k] = v }
	}
	r.data.Store(&registryData{definitions: newDefs, handlers: newHandlers})
	return nil
}

// GetDefinitions returns all tool definitions. Lock-free.
func (r *Registry) GetDefinitions() []kernel.ToolDefinition {
	d := r.load()
	defs := make([]kernel.ToolDefinition, 0, len(d.definitions))
	for _, v := range d.definitions { defs = append(defs, v) }
	return defs
}

// GetDefinitionsByNames returns selected tool definitions. Lock-free.
func (r *Registry) GetDefinitionsByNames(names []string) []kernel.ToolDefinition {
	d := r.load()
	defs := make([]kernel.ToolDefinition, 0, len(names))
	for _, name := range names {
		if def, ok := d.definitions[name]; ok { defs = append(defs, def) }
	}
	return defs
}

// Execute runs a tool call. Lock-free map lookup.
func (r *Registry) Execute(ctx context.Context, call kernel.ToolCall, sessionID string) (*kernel.ToolResult, error) {
	d := r.load()
	handler, ok := d.handlers[call.Function.Name]
	if !ok {
		return &kernel.ToolResult{Error: fmt.Sprintf("unknown tool: %s", call.Function.Name)}, nil
	}
	return handler(ctx, call.Function.Arguments)
}

// HasTool checks if a tool exists. Lock-free.
func (r *Registry) HasTool(name string) bool {
	_, ok := r.load().handlers[name]
	return ok
}

// Count returns the number of registered tools. Lock-free.
func (r *Registry) Count() int { return len(r.load().definitions) }

File: packages/tools/src/registry.ts
/**
 * 工具注册表 —— 内核 ToolExecutor 实现。
 * 线程无关（Node 单线程），工具按全名（含插件前缀）注册/分发/注销。
 */
import { ToolCall, ToolDefinition, ToolHandler, ToolResult, ToolExecutor } from '@openaide/core';

export class ToolRegistry implements ToolExecutor {
  private defs = new Map<string, ToolDefinition>();
  private handlers = new Map<string, ToolHandler>();

  register(def: ToolDefinition, handler: ToolHandler): void {
    const name = def.function.name;
    if (!name) throw new Error('tool name is empty');
    if (this.defs.has(name)) {
      throw new Error(`tool already registered: ${name}`);
    }
    this.defs.set(name, def);
    this.handlers.set(name, handler);
  }

  unregister(name: string): void {
    this.defs.delete(name);
    this.handlers.delete(name);
  }

  definitions(): ToolDefinition[] {
    return [...this.defs.values()];
  }

  getHandler(name: string): ToolHandler | undefined {
    return this.handlers.get(name);
  }

  async execute(toolCall: ToolCall, sessionId: string, signal?: AbortSignal): Promise<ToolResult> {
    const name = toolCall.function.name;
    const handler = this.handlers.get(name);
    if (!handler) {
      return { content: '', error: `unknown tool: ${name}`, errorCode: 'NOT_FOUND' };
    }
    try {
      return await handler(toolCall.function.arguments, sessionId, signal);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      const isTimeout = message.toLowerCase().includes('timeout');
      return {
        content: '',
        error: message,
        errorCode: isTimeout ? 'TIMEOUT' : 'EXEC_FAILED',
        isRetryable: isTimeout,
      };
    }
  }
}
